vdr/didx509/resolver.go

```diff
@@ -0,0 +1,232 @@
+		}
+		for i := 0; i < len(keyValue); i = i + 2 {
+			subject := cert.Subject
+			key := keyValue[i]
+			value := keyValue[i+1]
+			switch key {
+			case "serialNumber":
+				if subject.SerialNumber != value {
+					return ErrWrongSerialNumber
+				}
+			case "CN":
+				if subject.CommonName != value {
+					return ErrWrongCN
+				}
+			case "L":
+				if !slices.Contains(subject.Locality, value) {
+					return ErrWrongLocality
+				}
+			case "O":
+				if !slices.Contains(subject.Organization, value) {
+					return ErrWrongOrganization
+				}
+			case "OU":
+				if !slices.Contains(subject.OrganizationalUnit, value) {
+					return ErrWrongOrganizationalUnit
+				}
+			}
+		}
+	case "san":
+		keyValue := strings.Split(ref.PolicyValue, ":")
+		if len(keyValue)%2 != 0 {
+			return ErrDidSanMalformed
+		}
+		for i := 0; i < len(keyValue); i = i + 2 {
+			key := keyValue[i]
+			value := keyValue[i+1]
+			switch key {
+			case "otherName":
+				nameValue, err := findOtherNameValue(cert)
+				if err != nil {
+					return err
+				}
+				if nameValue != value {
+					return ErrWrongSanOtherName
+				}
+			case "dns":
+				if len(cert.DNSNames) > 0 && !slices.Contains(cert.DNSNames, value) {
+					return ErrWrongSanDns
+				}
+			case "email":
+				if len(cert.EmailAddresses) > 0 && !slices.Contains(cert.EmailAddresses, value) {
+					return ErrWrongSanEmailAddresses
+				}
+			case "ip":
+				if len(cert.IPAddresses) > 0 {
+					ok := false
+					for _, ip := range cert.IPAddresses {
+						if ip.String() == value {
+							ok = true
+							break
+						}
+					}
+					if !ok {
+						return ErrWrongSanIPAddresses
+					}
+
+				}
+			}
+
+		}
+	}
+
+	return nil
+}
+
+// createDidDocument generates a new DID Document based on the provided DID identifier and validation certificate.
+func createDidDocument(id did.DID, validationCert *x509.Certificate) (*did.Document, error) {
+	verificationMethod, err := did.NewVerificationMethod(did.DIDURL{DID: id}, ssi.JsonWebKey2020, id, validationCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	document := &did.Document{
+		Context: []interface{}{
+			ssi.MustParseURI("https://www.w3.org/ns/did/v1"),
+		},
+		ID:                 id,
+		Controller:         []did.DID{id},
+		VerificationMethod: did.VerificationMethods{verificationMethod},
+	}
+	document.AddAssertionMethod(verificationMethod)
+	document.AddCapabilityDelegation(verificationMethod)
+	document.AddCapabilityInvocation(verificationMethod)
+	document.AddKeyAgreement(verificationMethod)
+	return document, nil
+}
+
+// parseX509Did parses a DID (Decentralized Identifier) in the x509 format and returns a corresponding X509DidReference.
+func parseX509Did(id did.DID) (*X509DidReference, error) {
+	ref := X509DidReference{}
+	fullDidString := id.ID
+	policyString := ""
+	didString := ""
+	fullDidParts := strings.Split(fullDidString, "::")
+	if len(fullDidParts) == 1 {
+		didString = fullDidParts[0]
+	} else if len(fullDidParts) == 2 {
```

Comment:
missing test coverage

Reply:
Added tests, and PolicyNone. The latter will get removes as soon as we support multiple policies.